### PR TITLE
feat(dashboard): dashboard/id/datasets endpoint

### PR DESCRIPTION
**Pull request by suddjian**
### SUMMARY

Adding a new endpoint that gets the dataset metadata for a dashboard.

This is intended to replace the template-loaded bootstrap data on the dashboard page. In keeping with the precedent set by the existing bootstrap data, I have opted to return data in the same shape (a map of datasets) rather than a list, and to copy the bootstrap data's "trimming" behavior where any data unnecessary for rendering the dashboard is stripped out of the result. I don't believe these design decisions make for the ideal endpoint, but they do reduce potential regressions in the Single Page App project.

### TEST PLAN

Wrote unit tests, tested manually

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Changes UI
- [ ] Requires DB Migration.
- [ ] Confirm DB Migration upgrade and downgrade tested.
- [x] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

File: tests/dashboards/api_tests.py [hunk 1]
```diff
@@ -170,6 +170,31 @@ def create_dashboard_with_report(self):
...
+        dashboard = Dashboard.get("world_health")
+        dataset_ids = set([s.datasource_id for s in dashboard.slices])
+        self.assertEqual(len(data["result"]), len(dataset_ids))
+
+    @pytest.mark.usefixtures("load_world_bank_dashboard_with_slices")
+    def test_get_dashboard_datasets_not_found(self):
+        self.login(username="alpha")
+        uri = "api/v1/dashboard/not_found/datasets"
+        response = self.get_assert_metric(uri, "get_datasets")
+        self.assertEqual(response.status_code, 404)
```
Comment (amitmiran137) [inline tests/dashboards/api_tests.py:189]: could you also add security access tests to make sure dashboard_rbac enforces access all new dashboard sub objects endpoints?

here is a test I add for get_charts :
https://github.com/apache/superset/pull/13530

Comment (amitmiran137) [reply tests/dashboards/api_tests.py:189]: Ohhh it does satisfy, you're right thanks .

File: superset/dashboards/dao.py [hunk 2]
```diff
@@ -57,6 +58,29 @@ def get_by_id_or_slug(id_or_slug: str) -> Dashboard:
...
+            .outerjoin(Slice, Dashboard.slices)
+            .outerjoin(Slice.table)
+        )
+        # Apply dashboard base filters
+        query = DashboardFilter("id", SQLAInterface(Dashboard, db.session)).apply(
+            query, None
+        )
+        dashboard = query.one_or_none()
+        if not dashboard:
+            raise DashboardNotFoundError()
```
Comment (dpgaspar) [inline superset/dashboards/dao.py:75]: Feels kind of strange that a `DAO` raises a `Command` Exception

File: superset/dashboards/api.py [hunk 3]
```diff
@@ -252,6 +253,54 @@ def get(self, id_or_slug: str) -> Response:
...
+                  responses:
+                    200:
+                      description: Dashboard dataset definitions
+                      content:
+                        application/json:
+                          schema:
+                            type: object
+                            properties:
+                              result:
+                                type: object
```
Comment (dpgaspar) [inline superset/dashboards/api.py:287]: Better to declare a marshmallow schema and reference it here

**Review by dpgaspar — COMMENTED** [review]
Looks good left a couple of comments

File: superset/dashboards/dao.py (hunk 2 above)
Comment (suddjian) [reply superset/dashboards/dao.py:75]: Oh I never noticed that. It also feels kind of strange to create multiple different exception classes with the exact same meaning. I think it's reasonable to share them.

File: tests/dashboards/api_tests.py [hunk 4]
```diff
@@ -170,6 +170,31 @@ def create_dashboard_with_report(self):
...
+    @pytest.mark.usefixtures("load_world_bank_dashboard_with_slices")
+    def test_get_dashboard_datasets(self):
+        self.login(username="admin")
+        uri = "api/v1/dashboard/world_health/datasets"
+        response = self.get_assert_metric(uri, "get_datasets")
+        self.assertEqual(response.status_code, 200)
+        data = json.loads(response.data.decode("utf-8"))
+        dashboard = Dashboard.get("world_health")
+        dataset_ids = set([s.datasource_id for s in dashboard.slices])
+        self.assertEqual(len(data["result"]), len(dataset_ids))
```
Comment (villebro) [inline tests/dashboards/api_tests.py:182]: nit: we could probably compare a set of `id`s from `data["result"]` with `dataset_ids` here to make sure the ids are identical.

**Review by villebro — APPROVED** [review]
LGTM - just one minor improvement proposal for tests.

**Pull request merged**